### DISCUSSION TITLE
[JS101] 有關內建函式的疑問

### DISCUSSION THREAD

**Discussion by qoosuperman** [Lidemy/forum#61] [前端相關問題]
1. 有關 toString 這個 function，為什麼一定要把數字 assign 給變數之後才能使用呢? 如下：
```js
> 1.toString()
Thrown:
1.toString()
^^

> var a =1
> a.toString()
'1'
```
2. 請問為什麼有些內建函式的使用上，有一些都可以直接使用，或是前面可以加上另一個 function 都能正常運作呢？ 像是下面這樣：
```js
> console.log(Number.parseFloat('10.3'))
10.3
undefined
> console.log(parseFloat('10.3'))
10.3
undefined
> self
Thrown:
ReferenceError: self is not defined
> console.log(typeof(Number))
function
```
我之前學的語言是 Ruby，一個很物件導向的語言，所以我會覺得上面這個例子中，`parseFloat()` 這個 function 是屬於某個物件的，但用 console.log 印出來的東西表示 Javascript 顯然不是這樣子 XD

**Comment by aszx87410** [comment]
1. 因為語法的關係，`1.` 好像會被當作是 `1.0` 之類的，重點是那個點不會被解釋成 `1 上面的 method` 所以你要 `(1).toString()` 就可以了，或是很奇怪的 `1..toString()`
2.  兩個是同一個沒錯：`Number.parseFloat === parseFloat // true`，會可以直接呼叫是因為被放到 `window` 去，變成像是全域變數的東西，`window.parseFloat === Number.parseFloat // true`，然後他確實是屬於 Number

JS 很神奇，其實 function 也是一個物件，他在規格裡面被叫做是 invokable object 之類的（沒記錯的話），可以被呼叫的 object，所以其實底下程式碼是可以跑的：

``` js
function test(){}
test.a = 123
console.log(test.a) // 123
```

不知道 ruby 是不是也能這樣XD
